File: tests/tempHumSensor/custom/am2320.js
const fs = require('fs');
const path = require('path');

class AM2320 {
  constructor(busNumber = 1, address = 0x5C) {
    this.devicePath = `/dev/i2c-${busNumber}`;
    this.address = address;
    this.fd = null;
  }

  open() {
    if (this.fd) return;
    try {
      this.fd = fs.openSync(this.devicePath, 'r+');
      // Set the I2C slave address
      this._ioctl(0x0703, this.address);
      console.log(`Connected to AM2320 on ${this.devicePath} at address 0x${this.address.toString(16)}`);
    } catch (err) {
      throw new Error(`Failed to open I2C device: ${err.message}`);
    }
  }

  close() {
    if (this.fd !== null) {
      fs.closeSync(this.fd);
      this.fd = null;
    }
  }

  _ioctl(cmd, arg) {
    // This requires the native ioctl binding
    const { ioctl } = require('./ioctl_native.cc');
    return ioctl(this.fd, cmd, arg);
  }

  _writeBytes(bytes) {
    return fs.writeSync(this.fd, Buffer.from(bytes));
  }

  _readBytes(length) {
    const buffer = Buffer.alloc(length);
    fs.readSync(this.fd, buffer, 0, length, null);
    return buffer;
  }

  _sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  // Calculate CRC16 (Modbus)
  _calculateCRC16(buffer) {
    let crc = 0xFFFF;
    
    for (let i = 0; i < buffer.length; i++) {
      crc ^= buffer[i];
      
      for (let j = 0; j < 8; j++) {
        if (crc & 0x0001) {
          crc >>= 1;
          crc ^= 0xA001;
        } else {
          crc >>= 1;
        }
      }
    }
    
    return crc;
  }

  async read() {
    this.open();

    // Wake up the sensor (AM2320 requires this)
    try {
      this._writeBytes([0x00]);
    } catch (e) {
      // Expected to fail with EREMOTEIO, it's just to wake up the device
    }
    
    // Wait for sensor to wake up
    await this._sleep(10);
    
    // Send the read command - function code 0x03, start address 0x00, read 4 bytes
    this._writeBytes([0x03, 0x00, 0x04]);
    
    // Wait for the sensor to process
    await this._sleep(2);
    
    // Read the response (3 bytes of header + 4 bytes of data + 2 bytes of CRC)
    const response = this._readBytes(8);
    
    // Check function code
    if (response[0] !== 0x03) {
      throw new Error(`Invalid function code: ${response[0]}`);
    }
    
    // Check data length
    if (response[1] !== 0x04) {
      throw new Error(`Invalid data length: ${response[1]}`);
    }
    
    // Extract data
    const humidityHigh = response[2];
    const humidityLow = response[3];
    const temperatureHigh = response[4];
    const temperatureLow = response[5];
    
    // Calculate CRC
    const crc = (response[7] << 8) | response[6];
    const calculatedCRC = this._calculateCRC16(response.slice(0, 6));
    
    if (crc !== calculatedCRC) {
      throw new Error(`CRC check failed: ${crc} !== ${calculatedCRC}`);
    }
    
    // Calculate actual values
    const humidity = ((humidityHigh << 8) | humidityLow) / 10.0;
    const temperature = ((temperatureHigh << 8) | temperatureLow) / 10.0;
    
    return { temperature, humidity };
  }
}

// Export the AM2320 class
module.exports = AM2320;

File: telemetry/sensors/am2320.py
import smbus
import time

class AM2320:
    def __init__(self, bus_number=1, address=0x5C):
        """Initialize the AM2320 sensor.
        
        Args:
            bus_number (int): I2C bus number (default: 1)
            address (int): I2C address of the sensor (default: 0x5C)
        """
        self.bus = smbus.SMBus(bus_number)
        self.address = address
    
    def wake_up(self):
        """Wake up the sensor by sending a command."""
        try:
            self.bus.write_byte(self.address, 0x00)
        except:
            pass
        time.sleep(0.1)
    
    def read_data(self):
        """Read temperature and humidity data from the sensor.
        
        Returns:
            tuple: (temperature, humidity) or (None, None) if reading fails
        """
        try:
            # Wake up the sensor
            self.wake_up()
            
            # Send read command
            self.bus.write_i2c_block_data(self.address, 0x03, [0x00, 0x04])
            time.sleep(0.05)
            
            # Read data
            data = self.bus.read_i2c_block_data(self.address, 0x00, 8)
            
            # Calculate temperature and humidity
            humidity = (data[2] << 8 | data[3]) / 10.0
            temperature = (data[4] << 8 | data[5]) / 10.0
            
            return temperature, humidity
        except Exception as e:
            print(f"Error reading from AM2320: {e}")
            return None, None
    
    def get_temperature(self):
        """Get only the temperature reading.
        
        Returns:
            float: Temperature in Celsius or None if reading fails
        """
        temperature, _ = self.read_data()
        return temperature
    
    def get_humidity(self):
        """Get only the humidity reading.
        
        Returns:
            float: Humidity percentage or None if reading fails
        """
        _, humidity = self.read_data()
        return humidity


# Example usage
if __name__ == "__main__":
    sensor = AM2320()
    temperature, humidity = sensor.read_data()
    
    if temperature is not None and humidity is not None:
        print(f"Temperature: {temperature}°C")
        print(f"Humidity: {humidity}%")
    else:
        print("Failed to read sensor data")